### ISSUE TITLE
BLIT oscillators are a bit late

### ISSUE THREAD

**Issue by Andreya-Autumn** [surge-synthesizer/surge#7570]
Or should that be "a blit late"? :P

Anyway: 
![Old vs  new?](https://github.com/surge-synthesizer/surge/assets/105538426/994b8479-01f7-4696-8a4b-e504634f670f)

Classic seems to start late by a varying amount somewhere between 40 and 80 samples, wavetable ditto by about 80-160 samples. 

S&H Noise is interesting. With Correlation at max or min it behaves exactly like Classic. But with correlation at default it seems like it's accurate! As accurate as the MIDI block processing will let it that is. Maybe there's a clue there somehow?

**Comment by blancoberg** [comment]
I stumbled upon this bug as well. To me it looks like instead of starting on a random phase, it instead delays the signal for 0-1 cycle. So the delay varies depending on what note you are playing.
On lower notes the delay can be quite substantial

**Comment by mkruselj** [comment]
That's a plausible explanation!

**Comment by baconpaul** [comment]
Yes that’s exactly what they do and what the test above shows.

the fix to in init start them at start phase - 1 then run them up to zero but when I did that they clicked like crazy at low attacks.

**Comment by lluisestape-upc** [comment]
I had a dig at this, and I think the numbers in the original report pin the mechanism down exactly.

`oscstate` is measured in cycles, not samples — `process_block` subtracts `a = BLOCK_SIZE_OS * pitchmult` per block, and `pitchmult` is the reciprocal of the period. So in the non-retrigger branch of `ClassicOscillator::init`:

```cpp
double st = 0.5 * drand * storage->note_to_pitch_inv_tuningctr(detune);
oscstate[i] = st;
```

`oscstate` starts uniform in [0, 0.5) cycles. With `oscbuffer`/`dcbuffer` freshly memset, `last_level = 0` and `dc_uni = 0`, there's nothing to emit until the first `convolute()` fires — so the output is silent for `st * period` samples. That isn't a random start phase, it's a random start *delay*.

The reported numbers line up with that. At MIDI 60 / 44.1k the period is ~168 samples, so Classic should be silent for 0-84 samples; `WavetableOscillator::init` uses `oscstate[i] = drand` with no 0.5 factor, giving 0-168. That matches the "40 to 80" and "80 to 160" in the first post, including the 2x ratio between them.

If that's right, it would also explain why the `start phase - 1` and run-up approach clicked: it puts a real BLIT edge just after t=0, when the amp envelope is already open. Starting genuinely mid-cycle would instead mean priming the state that `init` currently zeroes — `last_level` set to the waveform level at the chosen phase, and `dc_uni` seeded into `dcbuffer[bufpos + FIRoffset]` so the ramp is already running at sample 0. No spurious edge, so nothing to click.

One thing I can't account for yet is the S&H observation that correlation at default is accurate while the extremes behave like Classic — `SampleAndHoldOscillator::init` uses the same [0, 0.5) seeding, so I'd expect it to be delayed too. I haven't worked out what's different there.

Two questions before I try anything:

1. This changes the render of every existing patch. Would you want it gated on streaming revision, or is a clean break fine here?
2. The `Oscillator Onset` regtest from #7571 computes `itUntilBigger` but doesn't currently assert on it. Would a real assertion with a per-oscillator threshold be welcome as part of the same PR?

**Comment by baconpaul** [comment]
Thanks for tackling this

Do you have an idea how much this changes sound? The click-at-onset is what i really worry about. Like try a few before and afters in the synth - anything worrying?

Chaging sample-for-sample render between versions is ok. changing sound materially is not. that's basically the stance we've taken. So if we can get some idea bout 'does this change the zero onset state materially' that would help. I think would be great to fix this in our current cycle if it really is just a slight onset of waveform change.

I'd also want to make sure that we didn't start in some cases with buffer which wasn't cleared etc.... would want to do an asan run for sure.

and yeah the onset test would want to assert to make sure ince it is fixed it stays fixed.

I'd definitely want it as a single commit we could bisect around though in case theres an unexpected side effect. Something like FM or phase coherence changing might (might) be a problem we would need to investigate in the future if it works.

**Comment by lluisestape-upc** [comment]
Thanks — that answers the gating question, I'll aim for a clean single commit with no streaming-revision branch, and treat "does it change the sound materially" as the bar to clear.

First, some numbers. I built the `Oscillator Onset` harness out into a distribution: 300 inits per oscillator, MIDI 60 at 44.1k, counting oversampled samples until the first `|output| > 1e-6`.

| Oscillator | retrig | min | max | mean |
|---|---|---|---|---|
| Classic | 1 | 1 | 1 | 1.0 |
| Classic | 0 | 1 | 169 | 86.5 |
| Wavetable | 1 | 2 | 2 | 2.0 |
| Wavetable | 0 | 3 | 335 | 158.5 |
| S&H Noise | 0 | 1 | 169 | 87.7 |
| Window / Sine / Twist / FM2 / FM3 | 0 | 1 | 1 | 1.0 |

The period at MIDI 60 / 44.1k is 337 oversampled samples. Classic seeds `oscstate` in [0, 0.5) cycles, predicting a 0-168 range and a mean of 84 for a uniform draw — measured 169 and 86.5. Wavetable seeds [0, 1), predicting 0-337 — measured 335. So the delay is `oscstate * period`, uniformly distributed.

Two things that came out of the sweep:

- The affected set is exactly the three `AbstractBlitOscillator` subclasses: Classic, Wavetable and S&H Noise. Window, Sine, Twist, FM2 and FM3 all sit at 1-2 samples regardless of retrigger, so the change only needs to touch those three `init()` methods. That should keep the FM and phase-coherence blast radius small, though I'll look at it explicitly.
- S&H Noise shows the same distribution as Classic here (max 169, mean 87.7) under `init_default_values()`, which doesn't match the "correlation at default seems accurate" note in the first post. Might be a harness-vs-GUI default difference — I'll pin that down rather than assume.

On the click worry: the reason I think this can avoid it is that the fix shouldn't emit an edge at all. Rather than running the oscillator up to phase zero, it primes the state `init()` currently zeroes — `last_level` to the waveform level at the chosen phase, and `dc_uni` seeded into `dcbuffer[bufpos + FIRoffset]` so the ramp is live from sample 0. The only discontinuity left is the initial level step, which is what retrigger mode already does at its edge. Whether that's audible at zero attack is exactly what needs A/B'ing, so that wants before/after renders across a spread of notes and attack times rather than an argument from theory.

**Comment by lluisestape-upc** [comment]
One more thing from reading the output stage, which I think explains the clicks from the `start phase - 1` approach.

`integrator_hpf` is `(1 - 2 * 20 * samplerate_inv)^2` — about 0.998187 at 44.1k — and it's applied once per oversampled sample in the `osc_out = osc_out * hpf + ...` loop. That decays to half in `ln(0.5) / ln(0.998187)`, roughly 382 oversampled samples, or about 1.1 cycles at MIDI 60.

Running the oscillator up from phase-1 to zero covers at most one cycle. So by the time the note actually starts, the leaky integrator hasn't settled from the initial edge that came out of silence — you begin the note carrying an un-decayed DC transient. That would make the click the transient rather than the edge itself, and it would get worse the lower the note, since the run-up is a fixed number of cycles but the decay is a fixed number of samples.

If that's right, it argues for priming the state analytically rather than running up to it: with no initial edge, there's nothing to settle from.

It is more state than I implied earlier, though. Besides `last_level` and `dc_uni` per voice, `osc_out` and the `dc` accumulator are per-oscillator rather than per-voice, so under unison they need to be seeded with the sum across voices rather than set per voice. The level within a segment is linear between the post-impulse level and `last_level`, so the arithmetic is tractable — but it does mean `init` needs to replicate part of the state machine in `convolute`. I'd rather factor that out into something both call than copy it, if you're open to that shape.

**Comment by Andreya-Autumn** [comment]
You mean higher notes get worse Right? The 384 samples have enough time to settle over one cycle at 50hz but not at 5000hz. Anyway agree seems we should set the state up correctly. 

The attack click thing might not be so bad. We're clicking anyway, It's just do we click when we're supposed to or slightly later. I bet there are situations where the difference is audible *if youre paying attention fot it* but I am much less sure if it's an audible change that anyone will ever actually miss if we do fix it.

**Comment by lluisestape-upc** [comment]
You're right, I had that backwards — higher notes are the bad case, not lower ones.

The run-up covers a fixed number of cycles while the integrator settles in a fixed number of samples, so at 50 Hz the run-up is ~1764 oversampled samples against a ~382 sample decay and there's plenty of room, while at 5 kHz it's ~18 samples against the same 382 and there's essentially none. So the argument for priming the state analytically holds, but the reason is that it removes a problem that bites in the treble.

Worth separating the two trends, since they run opposite ways:

- The onset delay itself is worst in the bass. It's a fixed fraction of a cycle, so in absolute time it grows as the note gets lower — at MIDI 60 it's up to 84 samples, and an octave down it's up to 168.
- The click from running up to phase zero is worst in the treble, for the settling reason above.

And agreed on the attack click. If we're producing an edge at note-on either way, the question is only whether it lands where it should or up to a cycle late, which is a much easier bar than "must be inaudible". That said I'd still like to get the before/afters in front of you both rather than assert it.

**Comment by Andreya-Autumn** [comment]
Sounds good!

**Comment by baconpaul** [comment]
Yeah indeed. let’s generate a PR and we can try it! When you do please also add yourself to the AUTHORS list.

On and since you appear to be using coding assistants (which is totally fine! This is exactly the type of issue they are great at finding too) our rough policy here is

- Please do a human review of code for surge before submitting
- Please run tests and try the synth to make sure its not ludicrous
- Please use the tag ‘Assisted-By: (your model name)’ as opposed to ‘Co-Authored-By’ or skipping model attribution.

We’ll review and test it also of course but the above steps make sure we aren’t reviewing stuff which folks never ran and submitted. I see no reason you believe you would do that but just thought it would be useful to share.
